src/vcf/sqlite_report.cpp

```diff
@@ -0,0 +1,310 @@
+        }
+
+        char *zErrMsg = NULL;
+        rc = sqlite3_exec(db, "CREATE TABLE if not exists errors (line int, code int, message varchar(255));", NULL, 0,
+                          &zErrMsg);
+        if (rc != SQLITE_OK) {
+            std::string error_message{std::string{"Can't use table: "} + zErrMsg};
+            sqlite3_free(zErrMsg);
+            sqlite3_close(db);
+            throw std::runtime_error{error_message};
+        }
+
+        rc = sqlite3_exec(db, "CREATE TABLE if not exists warnings (line int, code int, message varchar(255));", NULL, 0,
+                          &zErrMsg);
+        if (rc != SQLITE_OK) {
+            std::string error_message{std::string{"Can't use table: "} + zErrMsg};
+            sqlite3_free(zErrMsg);
+            sqlite3_close(db);
+            throw std::runtime_error{error_message};
+        }
+    }
+
+    void SqliteReportRW::flush()
+    {
+        // in case the amount of inserts is not a multiple of the transaction size, do the leftover inserts in a smaller transaction
+        if (current_transaction_size != 0) {
+            commit_transaction();
+            current_transaction_size = 0;
+        }
+    }
+    
+    void SqliteReportRW::close()
+    {
+        flush();
+        
+        if (db != nullptr) {
+            int rc = sqlite3_close(db);
+            if (rc != SQLITE_OK) {
+                int remaining_tries = 10;
+                while (rc == SQLITE_BUSY && remaining_tries-- > 0) {
+                    // maybe sqlite is still writing the last transaction
+                    std::cerr << "Waiting to finish transactions" << std::endl;
+                    std::this_thread::sleep_for(sleep_time);
+                    rc = sqlite3_close(db);
+                }
+                if (rc != SQLITE_OK) {
+                    throw std::runtime_error{"Can not close database " + db_name};
+                }
+            }
+            db = nullptr;
+        }
+    }
+
+    SqliteReportRW::~SqliteReportRW()
+    {
+        try {
+            close();
+        } catch (std::runtime_error e) {
+            std::cerr << "An error occurred finalizing the error reporting: " << e.what() << std::endl;
+        }
+    }
+
+    void SqliteReportRW::write_error(const Error &error)
+    {
+        write(error, "errors");
+    }
+
+    void SqliteReportRW::write_warning(const Error &error)
+    {
+        write(error, "warnings");
+    }
+
+    void SqliteReportRW::write(const Error &error, std::string table)
+    {
+        char *zErrMsg = NULL;
+        int rc;
+        if (current_transaction_size == 0) {
+            start_transaction();
+        }
+        std::stringstream ss;
+        ss << "INSERT INTO " << table << " ( line , code , message ) VALUES (";
+        ss << error.get_line() << " , "
+        << static_cast<int>(error.get_code()) << " , \""
+        << error.get_raw_message() << "\");";
+
+        rc = sqlite3_exec(db, ss.str().c_str(), NULL, 0, &zErrMsg);
```

Comment:
This makes SQLite3 compile a different statement every time an insert is executed, and types are not checked at the application level. https://www.sqlite.org/c3ref/bind_blob.html should be used to bind the parameters, and as a result `sqlite3_prepare_v2` instead of `sqlite3_exec`.